aptos-move/block-executor/src/errors.rs

```diff
@@ -4,17 +4,20 @@
 
 use aptos_aggregator::types::PanicOr;
 
-/// The same module access path for module was both read & written during speculative executions.
-/// This may trigger a race due to the Move-VM loader cache implementation, and mitigation requires
-/// aborting the parallel execution pipeline and falling back to the sequential execution.
-/// TODO: (short-mid term) relax the limitation, and (mid-long term) provide proper multi-versioning
-/// for code (like data) for the cache.
 #[derive(Copy, Clone, Debug, PartialEq, Eq)]
-pub struct ModulePathReadWrite;
+pub enum IntentionalFallbackToSequential {
+    /// The same module access path for module was both read & written during speculative executions.
+    /// This may trigger a race due to the Move-VM loader cache implementation, and mitigation requires
+    /// aborting the parallel execution pipeline and falling back to the sequential execution.
+    /// TODO: (short-mid term) relax the limitation, and (mid-long term) provide proper multi-versioning
+    /// for code (like data) for the cache.
+    ModulePathReadWrite,
+    DirectWriteSetTransaction,
+}
 
 #[derive(Clone, Debug, PartialEq, Eq)]
 pub enum Error<E> {
-    FallbackToSequential(PanicOr<ModulePathReadWrite>),
+    FallbackToSequential(PanicOr<IntentionalFallbackToSequential>),
```

Comment:
As `IntentionalFallbackToSequential` is an enum anyways, should we just add `CodeInvariantError` as a variant to the enum? Is PanicOr serving any additional benefit here?